kernel/inductive.ml

```diff
@@ -973,7 +973,7 @@ let rec subterm_specif renv stack t =
          let wf_args = (dest_subterms wf).(0) in
          (* We extract the tree of the projected argument *)
          let n = Projection.arg p in
-         spec_of_tree (List.nth wf_args n)
+         spec_of_tree wf_args.(n)
```

Comment:
This seems to be pretty much the only place where we use the O(1) indexing feature of arrays, and it's unlikely to matter algorithmically given that we keep converting between lists and arrays in the surrounding code. If possible, I think it would be good to seize the opportunity to replace the use of arrays in the Rtree API with lists. Arrays are definitely mutable in OCaml, and if not used for indexing, it's safer to provide an immutable API to the kernel so that nasty plugins cannot fiddle with the invariants of the structure, not mentioning potential issues with marshalling.

Reply:
As this is somehow orthogonal, I propose then to remove the second commit and that the immutability of the API is addressed on its own sake.

Reply:
All right.